src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,640 @@
+                        break;
+                    }
+                }
+                Err(err) => {
+                    debug!("{:?} Failed to receive request header: {:?}", self, err);
+                    break;
+                }
+            }
+        }
+    }
+
+    /// Get stream with mutable reference
+    #[allow(clippy::mut_from_ref)]
+    fn get_stream_mut(&self) -> &mut TcpStream {
+        // Current implementation is safe because the stream is only accessed by one thread
+        unsafe { &mut *self.stream.get() }
+    }
+
+    /// Get packet task with mutable reference
+    #[allow(clippy::mut_from_ref)]
+    fn get_packets_keeper_mut(&self) -> &mut PacketsKeeper<P> {
+        // Current implementation is safe because the packet task is only accessed by one thread
+        unsafe { &mut *self.packets_keeper.get() }
+    }
+}
+
+unsafe impl<P> Send for RpcClientConnectionInner<P> where P: Packet + Clone + Send + Sync + 'static {}
+
+unsafe impl<P> Sync for RpcClientConnectionInner<P> where P: Packet + Clone + Send + Sync + 'static {}
+
+/// The RPC client definition.
+#[derive(Debug)]
+pub struct RpcClient<P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    /// The timeout options for the client.
+    /// TODO:
+    #[allow(dead_code)]
+    timeout_options: ClientTimeoutOptions,
+    /// Inner connection for the client.
+    inner_connection: Arc<RpcClientConnectionInner<P>>,
+    /// Client ID for the connection
+    /// TODO:
+    #[allow(dead_code)]
+    client_id: AtomicU64,
+}
+
+impl<P> RpcClient<P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    /// Create a new RPC client.
+    ///
+    /// We don't manage the stream is dead or clean
+    /// The client will be closed if the keep alive message is not received in 100 times
+    /// When the stream is broken, the client will be closed, and we need to recreate a new client
+    pub fn new(stream: TcpStream, timeout_options: &ClientTimeoutOptions) -> Self {
+        let client_id = AtomicU64::new(0);
+        let inner_connection = RpcClientConnectionInner::new(
+            stream,
+            timeout_options,
+            client_id.load(std::sync::atomic::Ordering::Acquire),
+        );
+
+        Self {
+            timeout_options: timeout_options.clone(),
+            inner_connection: Arc::new(inner_connection),
+            client_id,
+        }
+    }
+
+    /// Start client receiver
+    pub fn start_recv(&self) {
+        // Create a receiver loop
+        let inner_connection_clone = Arc::clone(&self.inner_connection);
+        tokio::spawn(async move {
+            inner_connection_clone.recv_loop().await;
+        });
+    }
+
+    /// Send a request to the server.
+    /// Try to send data to channel, if the channel is full, return an error.
+    /// Contains the rezquest header and body.
+    /// WARN: this function does not support concurrent call
+    pub async fn send_request(&self, req: &mut P) -> Result<(), RpcError> {
```

Comment:
`req: &mut P` should change to `req: mut P`

Reply:
Ok

Reply:
Modified to `mut req: P`.